### PR TITLE
ts2pant: conditional mutations via symbolic last-write tracking

### PR DESCRIPTION
**Pull request by subsetpark**
## Summary

Translates conditional mutations in TypeScript function bodies into Pantagruel `cond` expressions, closing the gap that previously caused any `if`/`else` inside a mutating function to be rejected as `unsupported: "conditional assignment (if/else)"`.

The implementation is forward symbolic execution with path merging — Dijkstra's guarded commands (CACM 1975) for the encoding, Allen et al.'s if-conversion (POPL 1983) for the control-flow flattening, and Lamport's TLA+ frame treatment (*Specifying Systems* Ch. 2-3) for the identity-case handling in one-armed `if`s.

### Commits

1. **`9932c38` — symbolic last-write tracking.** A `SymbolicState` threaded through `translateBody*` records the last assigned value per mutable location keyed by `${prop}::${ast.strExpr(objExpr)}`. Property-access reads consult this state so later statements see prior writes. `if`/`else` clones state per branch, merges at the join via `cond(g => thenVal, true => elseVal)`, and falls back to the pre-state identity `prop obj` for untouched arms. Guards must be pure (same check as const initialisers).

2. **`f6e2e28` — early-exit if-conversion.** `if (g) { return; }` followed by statements lifts the continuation into a `cond` guarded by `!g`. Detection is limited to bare `return;` — richer early-exit shapes (return with a value, mixed side effects before return) remain unsupported.

3. **`5eaa7c1` — compound assigns + else-branch returns.** `a.p OP= v` desugars to `a.p = a.p OP v`, with the rhs's property-access read threading through the symbolic state. Early-exit detection extended to recognise `if (c) { X } else { return; }` as the symmetric form.

### Supported shapes

- `if (g) { obj.p = a }` → `p' obj = cond g => a, true => p obj`
- `if (g) { obj.p = a } else { obj.p = b }` → `p' obj = cond g => a, true => b`
- `obj.p = 1; if (g) { obj.p = 2 }` → `p' obj = cond g => 2, true => 1` (sequential composition)
- Nested `if`s, asymmetric writes (different props in each arm), read-modify-write
- `if (g) return; ... rest` (Allen if-conversion, early-exit extension)
- `if (g) { ... } else { return; }`
- Compound assigns: `+=`, `-=`, `*=`, `/=`, `%=`, `**=`

### Out of scope

- Loops (`for`, `while`, `do-while`) — require fixed-point or unrolling.
- `switch` — would extend if-conversion to multi-arm dispatch.
- `try/catch` with catch clause.
- Impure guards (calls with side effects in the if-condition) — rejected for soundness.
- Non-bare early exits (`return expr;`, mixed side effects before `return;`).

## Test plan

- [x] 9 snapshot fixtures in `functions-mutating-conditional.ts` covering single-arm, two-arm, nested, asymmetric, sequential, and read-modify-write cases.
- [x] 7 snapshot fixtures in `functions-mutating-early-exit.ts` covering the three early-exit patterns.
- [x] 5 snapshot fixtures in `functions-mutating-compound.ts` covering compound assigns (including sequential composition with prior writes).
- [x] 9 unit tests in `translate-body.test.mts` for the symbolic last-write edge cases.
- [x] End-to-end `pant --check` on `apply-fee.ts` verifies SMT acceptance of the emitted encoding.
- [x] Full test suite: 245 tests pass (up from 235 on master).

### CODE REVIEW

File: tools/ts2pant/src/translate-body.ts [hunk 1]
```diff
@@ -67,6 +67,124 @@ function bodyExpr(r: BodyResult): OpaqueExpr {
...
+    [ts.SyntaxKind.MinusEqualsToken, ts.SyntaxKind.MinusToken],
+    [ts.SyntaxKind.AsteriskEqualsToken, ts.SyntaxKind.AsteriskToken],
+    [ts.SyntaxKind.SlashEqualsToken, ts.SyntaxKind.SlashToken],
+    [ts.SyntaxKind.PercentEqualsToken, ts.SyntaxKind.PercentToken],
+    [
+      ts.SyntaxKind.AsteriskAsteriskEqualsToken,
+      ts.SyntaxKind.AsteriskAsteriskToken,
+    ],
+  ],
+);
```
[the remark this answers is not in the document]
Comment (subsetpark) [reply tools/ts2pant/src/translate-body.ts:199]: Good catch — fixed in 1da6d46. Removed `PercentEqualsToken` and `AsteriskAsteriskEqualsToken` from `COMPOUND_ASSIGN_TO_BINOP` (translate-body.ts). Pantagruel's AST has no `%` or `**` op (`ast.opMod`/`opPow` don't exist), so `translateOperator` would indeed reject the desugared RHS. Those tokens now fall through to the generic unsupported-operator path instead of pretending to be handled.

File: tools/ts2pant/src/translate-body.ts [hunk 2]
```diff
@@ -639,13 +768,21 @@ export function translateBodyExpr(
...
       }
     }
+    // Consult symbolic state for a prior write at this location.
+    if (state !== undefined) {
+      const key = symbolicKey(prop, bodyExpr(obj));
+      const entry = state.writes.get(key);
+      if (entry !== undefined) {
+        return { expr: entry.value };
+      }
+    }
```
[the remark this answers is not in the document]
Comment (subsetpark) [reply tools/ts2pant/src/translate-body.ts]: Fixed in 1da6d46. `SymbolicState` now carries a `canonicalize: (e) => OpaqueExpr` that mirrors the ambient `applyConst`, kept in sync whenever `symbolicExecute` composes a new const binding. The read site at translateBodyExpr:793 applies `state.canonicalize(bodyExpr(obj))` before building the lookup key, matching the post-`applyConst` write key at symbolicExecute:1491. Added fixture `aliasedSequentialWrites` (functions-mutating-const.ts) that exercises your exact example — snapshot confirms `balance' a = 1 + 2`, not `balance a + 2`.

File: tools/ts2pant/src/translate-body.ts [hunk 3]
```diff
@@ -1060,60 +1236,161 @@ function translateMutatingBody(
...
+        continuationBlock,
+        checker,
+        strategy,
+        new Map(paramNames),
+        sR,
+        remainingProps,
+        applyConst,
+        supply,
+        true,
+      );
```
[the remark this answers is not in the document]
Comment (subsetpark) [reply tools/ts2pant/src/translate-body.ts:1378]: Fixed in 1da6d46. Changed the `symbolicExecute` call for the continuation to pass `insideBranch` (the caller's value) instead of hard-coded `true`. The continuation lives at the caller's lexical scope, so it deserves the same early-exit-detection treatment. Added fixture `chainedEarlyReturns` (functions-mutating-early-exit.ts) — snapshot confirms the pattern `if (g) return; if (h) return; a.balance = 1;` now flattens to `balance' a = (cond g => balance a, true => (cond h => balance a, true => 1))`.

File: tools/ts2pant/tests/fixtures/constructs/functions-mutating-conditional.ts [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+  }
+}
+
+/** conditional branch reads an earlier unconditional write */
+export function accumulateIf(a: Account, g: boolean): void {
+  a.balance = 10;
+  if (g) {
+    a.balance = a.balance + 5;
+  }
+}
```
[the remark this answers is not in the document]
Comment (subsetpark) [reply tools/ts2pant/tests/fixtures/constructs/functions-mutating-conditional.ts:85]: Added in 1da6d46: `incrementIfPositiveCompound` and `accumulateIfCompound` in `functions-mutating-conditional.ts`, plus `aliasedSequentialWrites` in `functions-mutating-const.ts` (for the canonicalization fix) and `chainedEarlyReturns` in `functions-mutating-early-exit.ts` (for the continuation `insideBranch` fix). All four are auto-discovered by `constructs.test.mts` and the snapshots match the expected encodings.
